applications/PfemFluidDynamicsApplication/custom_elements/two_step_updated_lagrangian_V_P_implicit_fluid_FIC_cut_fem_element.cpp

```diff
@@ -0,0 +1,1413 @@
+        BaseType::CalculateLocalMomentumEquations(rLeftHandSideMatrix, rRightHandSideVector, rCurrentProcessInfo);
+
+        // If intersected, add the boundary contribution
+        if (IsCut())
+        {
+            // Calculate intersection Gauss points geometry data
+            Matrix interface_N;
+            ShapeFunctionDerivativesArrayType interface_DN_DX;
+            Vector interface_gauss_weights;
+            ModifiedShapeFunctions::AreaNormalsContainerType interface_unit_normals;
+            CalculateIntersectionGeometryData(interface_DN_DX, interface_N, interface_gauss_weights, interface_unit_normals);
+
+            // Create an auxiliary elemental data container for the boundary terms integration
+            ElementalVariables elemental_variables;
+            this->InitializeElementalVariables(elemental_variables);
+
+            // Get other data
+            const double kappa = rCurrentProcessInfo[PENALTY_COEFFICIENT];
+            KRATOS_ERROR_IF(kappa < 1.0e-12) << "'PENALTY_COEFFICIENT' is zero." << std::endl;
+            const double h = this->ElementSize();
+
+            // Check if the element is "slip"
+            // Note that if one node is flagged as SLIP the entire element is considered slip
+            bool is_slip = false;
+            const auto &r_geom = this->GetGeometry();
+            for (const auto& r_node : r_geom) {
+                if (r_node.Is(SLIP)) {
+                    is_slip = true;
+                    break;
+                }
+            }
+
+            // Interface Gauss points loop
+            array_1d<double,3> vel_gauss;
+            const double rho = this->mMaterialDensity;
+            const double dt = rCurrentProcessInfo[DELTA_TIME];
+            const double theta = this->GetThetaMomentum();
+            array_1d<double, TDim> proj_dev_stress;
+            const std::size_t n_nodes = r_geom.PointsNumber();
+            const unsigned int n_int_gauss_pts = interface_gauss_weights.size();
+            for (unsigned int g = 0; g < n_int_gauss_pts; g++)
+            {
+                // Get interface Gauss point data
+                const double g_weight = interface_gauss_weights[g];
+                const auto g_DN_DX = interface_DN_DX[g];
+                const auto g_shape_functions = row(interface_N, g);
+                const auto &r_g_unit_normal = interface_unit_normals[g];
+
+                // Calculate the mechanical response at the interface Gauss point to get the viscous stress
+                this->CalcMechanicsUpdated(elemental_variables, rCurrentProcessInfo, g_DN_DX);
+
+                auto &r_strain_vector = elemental_variables.SpatialDefRate;
+                auto &r_dev_stress_vector = elemental_variables.UpdatedDeviatoricCauchyStress;
+                auto &r_constitutive_matrix = elemental_variables.ConstitutiveMatrix;
+
+                auto p_cons_law = this->GetProperties().GetValue(CONSTITUTIVE_LAW);
```

Comment:
don't you  initialize the CL inside a `std::vector` like `mConstitutiveLawVector`?

Reply:
In the fluid as it is very rare to have history dependent materials we assume a unique constitutive law for the entire element, which is stored in an elemental member variable (rather than a vector). I think that this is extended to the PFEM fluid.

However, in here we are creating a copy from the properties' one each time we are accessing the Gauss point (most probably as a short cut solution while we were developing). This turns into a memory overhead that necessarily needs to be addressed. Thanks for spotting.

Reply:
See #11879. As a temporary solution I'd assign the pointer outside the Gauss point loop. Nevertheless, I think that this must be done at the `InitializeSolutionStep` (right now not possible).

Reply:
This should be solved now as the CutFEM element calls the `InitializeSolutionStep` of the base element which handles the pointer issue.